homeassistant/components/speedtestdotnet/__init__.py

```diff
@@ -3,83 +3,145 @@
-            }
-        )
-    },
-    extra=vol.ALLOW_EXTRA,
-)
-
 
 async def async_setup(hass, config):
+    """Intergation no longer supports importing from config."""
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
     """Set up the Speedtest.net component."""
-    conf = config[DOMAIN]
-    data = hass.data[DOMAIN] = SpeedtestData(hass, conf.get(CONF_SERVER_ID))
 
-    if not conf[CONF_MANUAL]:
-        async_track_time_interval(hass, data.update, conf[CONF_SCAN_INTERVAL])
+    client = SpeedTestClient(hass, config_entry)
 
-    def update(call=None):
-        """Service call to manually update the data."""
-        data.update()
+    if not await client.async_setup():
+        return False
 
-    hass.services.async_register(DOMAIN, "speedtest", update)
+    hass.data[DOMAIN] = client
 
-    hass.async_create_task(
-        async_load_platform(
-            hass, SENSOR_DOMAIN, DOMAIN, conf[CONF_MONITORED_CONDITIONS], config
-        )
-    )
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload SpeedTest Entry from config_entry."""
+    hass.services.async_remove(DOMAIN, SPEED_TEST_SERVICE)
+    if hass.data[DOMAIN].unsub_timer:
+        hass.data[DOMAIN].unsub_timer()
+
+    await hass.config_entries.async_forward_entry_unload(config_entry, "sensor")
+
+    hass.data.pop(DOMAIN)
 
     return True
 
 
-class SpeedtestData:
+class SpeedTestClient:
     """Get the latest data from speedtest.net."""
 
-    def __init__(self, hass, server_id):
+    def __init__(self, hass, config_entry):
         """Initialize the data object."""
+        self.hass = hass
+        self.config_entry = config_entry
         self.data = None
-        self._hass = hass
-        self._servers = [] if server_id is None else [server_id]
-
-    def update(self, now=None):
+        self.api = None
+        self.unsub_timer = None
+
+    @property
+    def scan_interval(self):
+        """Return the scan interval duration."""
+        if not self.config_entry.options.get(CONF_MANUAL):
+            return self.config_entry.options.get(
+                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+            )
+        return None
+
+    def update_data(self):
         """Get the latest data from speedtest.net."""
 
-        _LOGGER.debug("Executing speedtest.net speed test")
-        speed = speedtest.Speedtest()
-        speed.get_servers(self._servers)
-        speed.get_best_server()
-        speed.download()
-        speed.upload()
-        self.data = speed.results.dict()
-        dispatcher_send(self._hass, DATA_UPDATED)
+        server_id = self.get_server_id()
+        self.api.get_servers(servers=[server_id])
+
+        _LOGGER.debug(
+            "Executing speedtest.net speed test with server_id: %s", server_id
+        )
+        self.api.download()
+        self.api.upload()
+        self.data = self.api.results.dict()
+
+        async_dispatcher_send(self.hass, DATA_UPDATED)
```

Comment:
Please use our data update coordinator helper instead.

https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

Reply:
I am trying to implement the coordinator and just would like to know if the `scan_interval` can be configured as an option. In my testing, changing the scan_interval in options doesn't change the update_interval of the coordinator (already specified the scan_interval variable from the `entry.options[CONF_SCAN_INTERVAL]`).

Reply:
`update_interval` is a public attribute on the coordinator. Next time the coordinator updates, it will use the new interval, if the attribute is changed.

Reply:
I have tried to get this working but unfortunately `update_interval` is not updating its value when the config option `scan_interval` is updated. Can you please take a look at the last commit I pushed and let me know where I went wrong with this.